### ISSUE TITLE
Pooling support for high request rates?

### ISSUE THREAD

**Issue by jgimness** [JeringTech/Javascript.NodeJS#49]
Would there be problems having multiple instances?  Example use: A high traffic site that needs to do lots of vue/react SSR.  A single nodejs process could become CPU-limited.  

The readme says to avoid it, but are there any other consequences? 

> Try to avoid creating multiple INodeJSService instances since by default, each instance spawns a NodeJS process.

> Also, if you are using the static API from multiple threads and the NodeJS process is performing invocations for other threads, you might get unexpected results.

For the static API, does that mean we need to implement our own concurrency control on top of it? 

There's an older issue in the aspnet repo where someone put up some sample pooling code, for reference:
https://github.com/aspnet/JavaScriptServices/issues/1442#issuecomment-400963420

I think it would be so valuable to the .net community to have an out-of-the-box pooled nodejs service.

**Comment by JeremyTCD** [comment]
Good question, there's technically nothing wrong with multiple Node.js processes. A pool of INodeJSService instances would work fine. My concern was more with users repeatedly creating new instances, the ReadMe could be better worded. 

What are your thoughts on [Node.js workers](https://nodejs.org/docs/latest-v12.x/api/worker_threads.html) for parallelization of CPU intensive tasks? It's not out-of-the-box, but it'll have a much smaller memory footprint than multi-processing.

> For the static API, does that mean we need to implement our own concurrency control on top of it?

The static API is thread safe except for `Configure`:

> Configurations made using StaticNodeJSService.Configure<T> only apply to javascript invocations made using the static API. Ideally, such configurations should be done before the first javascript invocation. Any existing NodeJS process is killed and a new one is created in the first javascript invocation after every StaticNodeJSService.Configure<T> call. Re-creating the NodeJS process is resource intensive. Also, if you are using the static API from multiple threads and the NodeJS process is performing invocations for other threads, you might get unexpected results.

Basically, calling `Configure` restarts the underlying Node.js process (to apply the new options), so if multiple threads are using the underlying Node.js process when `Configure` is called, you'll likely get unexpected results.

**Comment by jgimness** [comment]
Thanks, I'll investigate Node.js workers and let you know what I find out.

I was going off research done by other projects, which seem to prefer separate processes (for isolation and redundancy), like here: https://medium.com/airbnb-engineering/operationalizing-node-js-for-server-side-rendering-c5ba718acfc9

Also ReactJS.NET does a similar model of pooling JS Engines (non-NodeJS, unfortunately) https://github.com/reactjs/React.NET/blob/3a8573a439be231cdd4cd702a6f4f425791b6e25/src/React.Core/JavaScriptEngineFactory.cs#L84

**Comment by JeremyTCD** [comment]
I've decided to add out-of-the-box parallelization #52 (work in progress). When it's done you'll just need to set `OutOfProcessNodeJSServiceOptions.Parallelization` to `Parallelization.MultiProcessing` or `Parallelization.MultiThreading`.

I think you may be right about separate processes being preferable - it seems there can be significant overhead passing data between threads using workers. We'll see once both have been implemented.  

Initial benchmarks for multi-processing are excellent:

```csharp
[Benchmark]
public async Task<DummyResult[]> INodeJSService_InvokeCpuBound_Multithreaded()
{
    const string dummyModule = @"module.exports = (callback, resultString) => {
// Block CPU
var end = new Date().getTime() + 1000;
while(new Date().getTime() < end){ /* do nothing */ } 

callback(null, {result: resultString});
}";
    const string dummyIdentifier = "dummyIdentifier";
    const string dummyResultString = "success";

    // Act
    const int numTasks = 5;
    var results = new Task<DummyResult>[numTasks];
    for (int i = 0; i < 5; i++)
    {
        results[i] = _nodeJSService.InvokeFromStringAsync<DummyResult>(dummyModule, dummyIdentifier, args: new[] { dummyResultString });
    }

    return await Task.WhenAll(results);
}
```

|                                      Method |    Mean |    Error |   StdDev | Gen 0 | Gen 1 | Gen 2 | Allocated |
|-------------------------------------------- |--------:|---------:|---------:|------:|------:|------:|----------:|
| INodeJSService_InvokeCpuBound_Multithreaded | 1.000 s | 0.0002 s | 0.0002 s |     - |     - |     - |  63.32 KB |

**Comment by JeremyTCD** [comment]
@jgimness #52 adds out-of-the-box multi-process concurrency. Documentation [here](https://github.com/JeringTech/Javascript.NodeJS#concurrency). Also cleaned up ReadMe issue you brought up. Kindly closet this issue if the solution works for you.

**Comment by DaniilSokolyuk** [comment]
@JeremyTCD  i tried concurency mode now and TryInvokeFromCacheAsync always false, i call that method about 1000 times and its return false always :(

**Comment by DaniilSokolyuk** [comment]
Ok, I see problem )
Maybe create method T InvokeFromCache(... Func<Stream> streamFacotry) overload?

**Comment by JeremyTCD** [comment]
@DaniilSokolyuk Oh yeah good catch, at present you end up caching in the "next" Node.js process in the pool. New overloads:

If module isn't cached in current Node.js process, factory is invoked and result is cached:
```
Task<T> InvokeFromCacheAsync<T>(string moduleCacheIdentifier, Func<Stream> moduleFactory, 
string exportName = null, object[] args = null, CancellationToken cancellationToken = default)
```
```
Task<T> InvokeFromCacheAsync<T>(string moduleCacheIdentifier, Func<string> moduleFactory, 
string exportName = null, object[] args = null, CancellationToken cancellationToken = default)
```
If module isn't cached in current Node.js process, module is sent to it and cached:
```
Task<T> InvokeFromCacheAsync<T>(string moduleCacheIdentifier, Stream moduleStream, 
string exportName = null, object[] args = null, CancellationToken cancellationToken = default)
```
```
Task<T> InvokeFromCacheAsync<T>(string moduleCacheIdentifier, string moduleString, 
string exportName = null, object[] args = null, CancellationToken cancellationToken = default)
```

At the same time I'm also thinking of adding overloads so args can be a `params` argument, been wanting to do this for awhile. E.g:

```
Task<T> InvokeFromCacheAsync<T>(string moduleCacheIdentifier, string moduleString, 
params object[] args)
```

Thoughts on these?

**Comment by JeremyTCD** [comment]
@DaniilSokolyuk Fix underway here: #57. Let me know if you have any input on the expanded API.

**Comment by JeremyTCD** [comment]
[Out-of-the-box concurrency documentation](https://github.com/JeringTech/Javascript.NodeJS#concurrency).
